contrib/codegen/src/database.rs

```diff
@@ -0,0 +1,144 @@
+    type_name: Ident,
+    /// The database name as passed in via #[database('database name')].
+    db_name: String,
+    /// The entire structure that the `database` attribute was called on.
+    structure: DataStruct,
+    /// The type inside the structure: struct MyDb(ThisType).
+    connection_type: Type,
+}
+
+const EXAMPLE: &str = "example: `struct MyDatabase(diesel::SqliteConnection);`";
+const ONLY_ON_STRUCTS_MSG: &str = "`database` attribute can only be used on structs";
+const ONLY_UNNAMED_FIELDS: &str = "`database` attribute can only be applied to structs with \
+    exactly one unnamed field";
+const NO_GENERIC_STRUCTS: &str = "`database` attribute cannot be applied to a struct with a \
+    generic type";
+
+const UNABLE_TO_APPLY: &str = "unable to apply the `database` attribute to the \
+    given struct";
+
+fn parse_invocation(attr: TokenStream, input: TokenStream) -> Result<DatabaseInvocation> {
+    let attr_stream2 = ::proc_macro2::TokenStream::from(attr);
+    let attr_span = attr_stream2.span();
+    let string_lit = ::syn::parse2::<LitStr>(attr_stream2)
+        .map_err(|_| attr_span.error("expected string literal"))?;
+
+    let input = ::syn::parse::<DeriveInput>(input).unwrap();
+
+    if !input.generics.params.is_empty() {
+        return Err(input.span().error(NO_GENERIC_STRUCTS));
+    }
+
+    let structure = match input.data {
+        Data::Struct(s) => s,
+        _ => return Err(input.span().error(ONLY_ON_STRUCTS_MSG))
+    };
+
+    let inner_type = match structure.fields {
+        Fields::Unnamed(ref fields) if fields.unnamed.len() == 1 => {
+            let first = fields.unnamed.first().expect("checked length");
+            first.value().ty.clone()
+        }
+        _ => return Err(structure.fields.span().error(ONLY_UNNAMED_FIELDS).help(EXAMPLE))
+    };
+
+    Ok(DatabaseInvocation {
+        type_name: input.ident,
+        db_name: string_lit.value(),
+        structure: structure,
+        connection_type: inner_type,
+    })
+}
+
+pub fn database_attr(attr: TokenStream, input: TokenStream) -> Result<TokenStream> {
+    let invocation = parse_invocation(attr, input)?;
+
+    if let Type::Path(type_path) = invocation.connection_type {
+        let inner_type = &type_path.path.segments.iter()
+            .map(|ps| ps.ident.clone())
+            .collect::<Vec<Ident>>();
+
+        let database_name = &invocation.db_name;
+        let handler_type = &invocation.type_name;
+        let pool_type = Ident::new(&format!("{}Pool", handler_type), Span::call_site());
+
+        let tokens = quote! {
+            use ::std::ops::*;
+            use ::rocket::http::*;
+            use ::rocket::request::*;
+            use ::rocket::*;
+            use ::rocket::fairing::*;
+            use ::rocket_contrib::*;
+            use ::rocket_contrib::databases::*;
+            pub struct #handler_type(r2d2::PooledConnection<<#(#inner_type)::* as Poolable>::Manager>);
```

Comment:
`r2d2` wasn't imported by the generated code, so this will break whenever it's used outside of the crate root.

If we had real proc-macros, we'd do `extern crate r2d2;` followed by `self::r2d2::ThingWeNeed`, but the best we can do here is assume `r2d2` is in the crate root and use `::r2d2`.

Reply:
Yeah, this makes sense. I'll adjust it to assume it was brought in inside the crate root.